markdown.go

```diff
@@ -218,7 +218,8 @@ type parser struct {
 	// Footnotes need to be ordered as well as available to quickly check for
 	// presence. If a ref is also a footnote, it's stored both in refs and here
 	// in notes. Slice is nil if footnotes not enabled.
-	notes []*reference
+	notes       []*reference
+	notesRecord map[string]bool
```

Comment:
1. About the `map[string]struct{}`, my understanding is that we can not use a map to represent footnotes because they have to be ordered, which means a slice is the best way like it is now `p.notes`. Another map variable `notesRecord` here is just to make the check of duplicate faster.

2. About `p.refs`, do you mean to use `p.refs` to check the duplicate? I do not think it works. As the comment above the definition of `notes []*refrence` says, `p.refs` represents all the references including image, footnote, url etc., while `p.notes` only contains _valid_ and sorted footnotes. It only makes sense to check `p.notes` because it is used to generate footer.

For example, when the input is
```
This uses footnote A.[^A]
                                                                                                                           
This uses footnote C.[^C]                                                                                                                                   
[^A]:  A note. use itself.[^A]                                                       
[^B]:  B note, uses A to test duplicate.[^A]                                         
[^C]:  C note, uses B.[^B]
[^D]: D note.
```
`p.refs` contains all four footnotes. But `p.notes` only contains A and C before parsing footers. And then when parse footnote C, the ref B is already in `p.refs` but not in `p.notes`. B should be added into `p.notes`. But if according to `p.refs`, it is the duplicate ref and not be added into `p.notes`, which is a wrong behavior.